### ISSUE TITLE
Cusync sync method problem

### ISSUE THREAD

**Issue by ziyuhuang123** [parasailteam/cusync#5]
https://github.com/parasailteam/cusync/blob/1a1375cba0f4620249e7974a0d00bd04205584ca/src/include/cusync/cusync_device_defs.h#L52-L68C1

Hi! I am learning your code, and I think here exists a ... strange sync method. Like the official example, https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#memory-fence-functions
Here we first compute the data, and then if(thread==0){ and then threadfence(), and then } and then syncthread(). But in the cusync, I see first we compute the data, and then syncthread()(?????), and then if(thread==0){ and then threadfence_system(is this really needed???) and then } and then syncwarp(???)

I guess... maybe here can be somehow simplified like the official guide? 

Any idea? Thanks!!!

**Comment by ziyuhuang123** [comment]
By the way, I changed the code like I guessed, it works on two matmul example. (just, yet....)

**Comment by abhijangda** [comment]
syncthreads is needed to ensure that before the semaphore is post (using atomicAdd) all threads have actually done the computation and memory writes have been issued.
Threadfence is needed to ensure that semaphore write is done after all memory writes are completed. Since only first thread is doing the atomicAdd,  we only need to do threadfence by first thread. 

syncwarp is needed to ensure that all diverged threads reconverge because of independent thread scheduling (see https://developer.nvidia.com/blog/inside-volta/ ).

Thank you reporting, I will add comments to explain this. I also recommend reading the paper: https://arxiv.org/abs/2305.13450 . 

Yes, threadfence is not needed when using cusync to synchronize kernels on the same GPU. So, it can be avoided when running kernels on the same GPU.

**Comment by ziyuhuang123** [comment]
For syncwarp, I guess you mean this?
![image](https://github.com/parasailteam/cusync/assets/99854690/7c8c60b2-d981-40b2-83fa-3a89476b0d30)
I think differently, like, syncwarp can only be within a warp, we are working on thread 0, but actually other threads not in warp0 are also waiting for thread 0's result. I guess syncthread is more suitable here?

**Comment by abhijangda** [comment]
Yes, syncwarp is needed only for the first warp  because the condition is only for thread 0. Indeed if a threadblock computes multiple tiles then syncthreads is more suitable.
